icu4j/main/classes/core/src/com/ibm/icu/message2/Mf2DataModelFormatter.java

```diff
@@ -0,0 +1,277 @@
+                FormattedValue fv = fr.getFormattedValue();
+                if (fr.getInput() == null && (fv == null || fv.toString() == null)) {
+                    result.append("{" + ((Expression) part).getOperand() + "}");
+                } else {
+                    result.append(fr.toString());
+                }
+            } else {
+                throw new IllegalArgumentException("Unknown part type: " + part);
+            }
+        }
+        return result.toString();
+    }
+
+    private Object variableToObjectEx(Value value, Map<String, Object> arguments) {
+        return variableToObjectEx(value, arguments, 0);
+    }
+
+    /*
+     * Pass a level to prevent local variables calling each-other recursively:
+     *
+     * <code><pre>
+     * let $l1 = {$l4 :number}
+     * let $l2 = {$l1 :number}
+     * let $l3 = {$l2 :number}
+     * let $l4 = {$l3 :number}
+     * </pre></code>
+     *
+     * We can keep track of the calls (complicated and expensive).
+     * Or we can forbid the use of variables before they are declared, but that is not in the spec (yet?).
+     */
+    private Object variableToObjectEx(Value value, Map<String, Object> arguments, int level) {
+        if (value == null) { // function only
+            return null;
+        }
+        // We have an operand. Can be literal, local var, or argument.
+        if (value.isLiteral()) {
+            return value.getLiteral();
+        } else if (value.isVariable()) {
+            String varName = value.getVariableName();
+            Expression localPh = dm.getLocalVariables().get(varName);
+            if (localPh != null && level < 5) {
+                return formatPlaceholder(localPh, arguments, false);
+            }
+            return arguments.get(varName);
+        } else {
+            throw new IllegalArgumentException("Invalid operand type " + value);
+        }
+    }
+
+    private FormattedPlaceholder formatPlaceholder(Expression ph, Map<String, Object> arguments, boolean localExpression) {
+        Object toFormat;
+        Value operand = ph.getOperand();
+        if (operand == null) { // function only, "...{:currentOs option=value}..."
+            toFormat = null;
+        } else {
+            // We have an operand. Can be literal, local var, or argument.
+            if (operand.isLiteral()) { // "...{(1234.56) :number}..."
+                // If it is a literal, return the string itself
+                toFormat = operand.getLiteral();
+            } else if (operand.isVariable()) {
+                String varName = operand.getVariableName();
+                if (!localExpression) {
+                    Expression localPh = dm.getLocalVariables().get(varName);
+                    if (localPh != null) {
+                        // If it is a local variable, we need to format that (recursive)
+                        return formatPlaceholder(localPh, arguments, true);
```

Comment:
I am surprised that the local variables aren't evaluated immediately when they are defined, before the actual pattern is formatted. Could you please add a TODO to think about it once more?

Reply:
Done.
Added TODO.

Reply:
Discussed: Evaluating local variables (an unfortunate name) immediately might waste time because not all of them may be used by the selected pattern. Lazy-once eval should be possible. The ICU User Guide (future chapter about MF2) and MFWG spec should say that the time of evaluation may vary. This might be observable if a variable expression yields an error when evaluated, which means that depending on the strategy some implementation may throw an error and another might ignore a locale variable.

Reply:
Created issue: https://github.com/unicode-org/message-format-wg/issues/299